File: LargestStack/largeststack.py
class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if len(self.items) > 0:
            return self.items.pop()
        else:
            return None
    
    def peek(self):
        if len(self.items) > 0:
            return self.items[len(self.items) - 1]
        else:
            return None

class MaxStack:
    def __init__(self):
        self.stack = Stack()
        self.maxes = Stack()

    def push(self, item):
        self.stack.push(item)
        # Check the latest max and compare values (if bigger, push)
        currentMax = self.maxes.peek()
        if currentMax is None or item >= currentMax:
            self.maxes.push(item)

    def pop(self):
        item = self.stack.pop()
        # See if popped item maxes latest max (if so, remove)
        if item == self.maxes.peek():
            self.maxes.pop()
        return item

    def getMax(self):
        return self.maxes.peek()

maxStack = MaxStack()
print(maxStack.getMax())   # should print None

maxStack.push(1)
print(maxStack.getMax())   # should print 1

maxStack.push(100)
print(maxStack.getMax())   # should print 100

maxStack.pop()
print(maxStack.getMax())   # should print 1

File: LargestStack/largeststack.js
class Stack {
    constructor() {
        this.items = [];
    }

    push(item) {
        this.items.push(item);
    }

    pop() {
        if (this.items.length) return this.items.pop();
        return null;
    }

    peek() {
        if (this.items.length) {
            return this.items[this.items.length - 1];
        }
        return null;
    }
}

// Goal is to find largest item in the stack
// So we create two stacks: first = all, second = max
class MaxStack {
    constructor() {
        this.stack = new Stack();
        this.maxes = new Stack();
    }

    push(item) {
        this.stack.push(item);
        // If given item > latest max (or no max exists), add as new max to maxes stack
        if (!this.maxes.peek() || item >= this.maxes.peek()) {
            this.maxes.push(item);
        }
    }

    pop() {
        const item = this.stack.pop();
        // If popped item is current max, remove previous max
        if (item === this.maxes.peek()) {
            this.maxes.pop();
        }
    }

    // Simply call peek, already defined
    getMax() {
        return this.maxes.peek();
    }
}

/* Some console.log tests */
const maxStack = new MaxStack();
console.log(maxStack.getMax());   // should print `null`

maxStack.push(1);
console.log(maxStack.getMax());   // should print 1

maxStack.push(100);
console.log(maxStack.getMax());   // should print 100

maxStack.pop();
console.log(maxStack.getMax());   // should print 1